fl4health/losses/deep_mmd_loss.py

```diff
@@ -0,0 +1,224 @@
+        Kxyxy = torch.cat((Kxxy, Kyxy), 0)
+        nx = Kx.shape[0]
+        ny = Ky.shape[0]
+        is_unbiased = True
+        if is_unbiased:
+            xx = torch.div((torch.sum(Kx) - torch.sum(torch.diag(Kx))), (nx * (nx - 1)))
+            yy = torch.div((torch.sum(Ky) - torch.sum(torch.diag(Ky))), (ny * (ny - 1)))
+            # one-sample U-statistic.
+            if use_1sample_U:
+                xy = torch.div((torch.sum(Kxy) - torch.sum(torch.diag(Kxy))), (nx * (ny - 1)))
+            else:
+                xy = torch.div(torch.sum(Kxy), (nx * ny))
+            mmd2 = xx - 2 * xy + yy
+        else:
+            xx = torch.div((torch.sum(Kx)), (nx * nx))
+            yy = torch.div((torch.sum(Ky)), (ny * ny))
+            # one-sample U-statistic.
+            if use_1sample_U:
+                xy = torch.div((torch.sum(Kxy)), (nx * ny))
+            else:
+                xy = torch.div(torch.sum(Kxy), (nx * ny))
+            mmd2 = xx - 2 * xy + yy
+        if not is_var_computed:
+            return mmd2, None, Kxyxy
+        hh = Kx + Ky - Kxy - Kxy.transpose(0, 1)
+        V1 = torch.dot(hh.sum(1) / ny, hh.sum(1) / ny) / ny
+        V2 = (hh).sum() / (nx) / nx
+        varEst = 4 * (V1 - V2**2)
+        return mmd2, varEst, Kxyxy
+
+    def MMDu(
+        self,
+        Fea: torch.Tensor,
+        len_s: int,
+        Fea_org: torch.Tensor,
+        sigma: torch.Tensor,
+        sigma0: torch.Tensor,
+        epsilon: torch.Tensor,
+        is_smooth: bool = True,
+        is_var_computed: bool = True,
+        use_1sample_U: bool = True,
+    ) -> Tuple[torch.Tensor, Optional[torch.Tensor], torch.Tensor]:
+        """compute value of deep-kernel MMD and std of deep-kernel MMD using merged data."""
+        X = Fea[0:len_s, :]  # fetch the sample 1 (features of deep networks)
+        Y = Fea[len_s:, :]  # fetch the sample 2 (features of deep networks)
+        X_org = Fea_org[0:len_s, :]  # fetch the original sample 1
+        Y_org = Fea_org[len_s:, :]  # fetch the original sample 2
+        L = 1  # generalized Gaussian (if L>1)
+        Dxx = self.Pdist2(X, X)
+        Dyy = self.Pdist2(Y, Y)
+        Dxy = self.Pdist2(X, Y)
+        Dxx_org = self.Pdist2(X_org, X_org)
+        Dyy_org = self.Pdist2(Y_org, Y_org)
+        Dxy_org = self.Pdist2(X_org, Y_org)
+        if is_smooth:
+            Kx = (1 - epsilon) * torch.exp(-((Dxx / sigma0) ** L) - Dxx_org / sigma) + epsilon * torch.exp(
+                -Dxx_org / sigma
+            )
+            Ky = (1 - epsilon) * torch.exp(-((Dyy / sigma0) ** L) - Dyy_org / sigma) + epsilon * torch.exp(
+                -Dyy_org / sigma
+            )
+            Kxy = (1 - epsilon) * torch.exp(-((Dxy / sigma0) ** L) - Dxy_org / sigma) + epsilon * torch.exp(
+                -Dxy_org / sigma
+            )
+        else:
+            Kx = torch.exp(-Dxx / sigma0)
+            Ky = torch.exp(-Dyy / sigma0)
+            Kxy = torch.exp(-Dxy / sigma0)
+
```

Comment:
Maybe here just add a comment noting what Kx, Ky, Kxy represent. That is, they are $k_w(x_i, x_j)$, $k_w(y_i, y_j)$ and $k_w(x_i,y_j)$ for all i, j in the samples of X and Y

Reply:
Perhaps we can also name these to be a bit more verbose. That is, `Kx` -> `kernel_xx` `K_y` -> `kernel_yy` and `K_xy` -> `kernel_xy`. We could do the same for `Dx`, `Dy`, `Dxy` -> `distance_xx`, `distance_yy`, `distance_xy`

Reply:
Done